pkg/planner/core/plan_cache_instance.go

```diff
@@ -0,0 +1,219 @@
+// Package core Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util/kvcache"
+	"go.uber.org/atomic"
+)
+
+// InstancePlanCache represents the instance/node level plan cache.
+type InstancePlanCache interface {
+	// Get gets the cached value from the cache according to key and opts.
+	Get(sctx sessionctx.Context, key string, opts any) (value any, ok bool)
+	// Put puts the key and value into the cache.
+	Put(sctx sessionctx.Context, key string, value, opts any) (succ bool)
+	// Evict evicts some cached values.
+	Evict(sctx sessionctx.Context) (evicted bool)
+	// MemUsage returns the total memory usage of this plan cache.
+	MemUsage(sctx sessionctx.Context) int64
+}
+
+// NewInstancePlanCache creates a new instance level plan cache.
+func NewInstancePlanCache(softMemLimit, hardMemLimit int64) InstancePlanCache {
+	planCache := new(instancePlanCache)
+	planCache.softMemLimit.Store(softMemLimit)
+	planCache.hardMemLimit.Store(hardMemLimit)
+	return planCache
+}
+
+type instancePCNode struct {
+	value    any // the underlying value, which should be (*PlanCacheValue)
```

Comment:
value can be defined as ```*PlanCacheValue```. only when to ```createNode```, we should make ```kvcache.Value``` as ```*PlanCacheValue```

Reply:
updated, PTAL